controller/proxy-injector/webhook.go

```diff
@@ -74,9 +77,29 @@ func Inject(
 		}
 		ownerKind = strings.ToLower(ownerRef.Kind)
 	}
+
+	configLabels := configToPrometheusLabels(resourceConfig)
 	proxyInjectionAdmissionRequests.With(admissionRequestLabels(ownerKind, request.Namespace, report.InjectAnnotationAt, configLabels)).Inc()
 
-	if injectable, reasons := report.Injectable(); !injectable {
+	// If a namespace has the opaque ports annotation, the resource should
+	// inherit it if it does not already have one itself. An initial patch is
+	// created if there is an opaque annotation that is not already on the
+	// workload.
+	var patchJSON []byte
+	annotation, onWorkload := resourceConfig.GetOpaquePorts()
+	if annotation != "" && !onWorkload {
```

Comment:
Would it make more sense to do this check after the injectable case?  I think that would let us use more early returns to simplify the logic.  i.e.

1. If it's an injectable pod, build the pod patch and return it
2. If it should have opaque ports copied, build an annotation patch and return it
3. otherwise, skip it (return an admission with no patch)

Note that there is some common logic at the bottom of this method for building the admission response, logging, and incrementing metrics that could get refactored into a function since it would be needed in each branch where we return a response with a patch.

Reply:
I responded above about the convenience of only checking `if annotation != "" && !onWorkload` once. It needs to be checked before the `injectable()` check because the opaque ports annotation needs to have been added before that.

We would need to check it again after the `injectable()` check in order to decide if an annotation patch should be created.

I'm not sure it really simplifies that much, it's just another order of checking that would work.

Reply:
Ah yeah, I hadn't realized how the annotation was used in both cases.  Still, I think the use of early returns is still valuable.  Something like:

```
opaquePorts, opaquePortsOk := resourceConfig.GetOpaquePorts()
if opaquePortsOk {
  resourceConfig.AppendPodAnnotation(pkgK8s.ProxyOpaquePortsAnnotation, opaquePorts)
}

// Case 1: Injectable
injectable, reasons := report.Injectable()
if !injectable && resourceConfig.IsPod() {
  // Log, record event, and track metrics for the reasons injection was skipped
}
if injectable {
  // [inject sidecar]
  return response, nil
}

// case 2: Annotate only
if opaquePortsOk {
  patchJSON, err = resourceConfig.CreateAnnotationPatch(annotation)
  // [annotate]
  return response, nil
}

// case 3: Just admit it
return response, nil
```